flowjax/train/variational_fit.py

```diff
@@ -0,0 +1,89 @@
+from typing import Optional, Callable
+
+from jax import random
+import optax
+import equinox as eqx
+from tqdm import tqdm
+
+from flowjax.utils import Array
+from flowjax.distributions import Distribution
+
+# Variational targets accept an array of proposed samples and return an array of (usually)
+# the unormalised log posterior probability of the samples.
+VariationalTarget = Callable[[Array], Array]
+
+# Variational losses are functions that take a distribution, a target 
+# callable and a random key, and returns a scalar loss.
+VariationalLoss = Callable[
+    [Distribution, VariationalTarget, random.KeyArray], 
+    float
+]
+
+@eqx.filter_jit
+def elbo_loss(dist: Distribution, target: VariationalTarget, key: random.KeyArray, elbo_samples: int = 500):
+    samples = dist.sample(key, sample_shape=(elbo_samples,))
+    approx_density = dist.log_prob(samples).reshape(-1)
+    target_density = target(samples).reshape(-1)
+    losses = approx_density - target_density
+    return losses.mean()
+
+def variational_fit(
+    key: random.KeyArray,
+    dist: Distribution,
+    target: VariationalTarget,
+    loss_fcn: VariationalLoss = elbo_loss,
+    num_epochs: int = 100,
+    optimizer: Optional[optax.GradientTransformation] = None,
+    show_progress: bool = True,
+    recorder: Optional[Callable] = None,
```

Comment:
This isn't documented. It seems to store the distribution at each step? I think I wouldn't bother having that unless there is a good reason for it.

Reply:
Yeah this is something I find useful when running small experiments to track specific behaviour over iterations.
Will strip it out and consider if I want to make an issue to add it.

Reply:
(obviously we can deal with that at a different point)

Reply:
Yeah, I agree that would be a good solution.